### ISSUE TITLE
Category tree is slow with many category/product associations.

### ISSUE THREAD

**Issue by romainruaud** [magento/magento2#7469]
### Preconditions

1. Magento 2.1.1
2. PHP 7 - Percona SQL Server on a stock production server.

### Steps to reproduce

1. Have 3.000.000 of product into your catalog
2. Have 6000 categories from level 1 to 8
3. Products are dispatched into the categories.
4. All categories are set to have is_anchor = 1.
5. Some top level categories are having more than 1.000.000 products associated

### Expected result

1. Display Time for Category Edit page should be reasonable (2-3 s)

### Actual result

1. Display time is around 10-15 s
2. => BO operations are slow, since it takes 10-15 secs to display a category edit page, and then 10-15 secs to save it (the save itself is quite fast, but the re-display of the page is slow)

I found that it seems to be mainly related to the category tree displaying. I can observe the following slow queries : 

```
SELECT COUNT(DISTINCT main_table.product_id) FROM `catalog_category_product` AS `main_table`  INNER JOIN `catalog_category_entity` AS `e` ON main_table.category_id=e.entity_id  WHERE (e.entity_id = '2') OR (e.path LIKE '1/2/%');
```

This one takes more than 6 seconds to run on my server (and returns 2.980.000 products), so it is an usual suspect !

**Comment by kandy** [comment]
Hi @romainruaud. Can I ask you to describe information about your database server, like memory amount, size of database in byte, innodb_buffer_pool_size?

Do you really need is_anchor = 1 setting for all categories?

**Comment by romainruaud** [comment]
Hello,

Hello @kandy and thank you for answering,

Server specs are the following : 

Processorr : Intel Xeon E5-2430@2.00Ghz * 2
RAM : 24 GB
Storage : 600GB x2 (RAID-1).
OS : Debian 8,5 (Jessie).

The innodb_buffer_pool_size is set to 16Go.

The Database size is approx. 30Go.

And yes, I need the is_anchor to 1 for all categories.

**Comment by naydav** [comment]
Thanks
We've created internal ticket MAGETWO-63383 to address this issue.

**Comment by jonpday** [comment]
any progress on this @naydav @miakusha @kandy ?

**Comment by choukalos** [comment]
Hi @romainruaud,

Wait - Do you have top level categories that have 1M products associated with them?  If so, wouldn't that make it very hard for a customer to browse the catalog and checkout?  

Out of curiosity how many websites / store views & customer groups do you have? 

Thanks,
Chuck

**Comment by veloraven** [comment]
This issue was already fixed in develop branch.
Internal ticket for ver. 2.1 - MAGETWO-65467

**Comment by george-vlahakis** [comment]
On version 2.1.7 now, is this meant to be fixed here @veloraven ?

**Comment by magento-engcom-team** [comment]
The issue already fixed in 2.2.0 
@romainruaud thank you for report

**Comment by psolutions-me** [comment]
Have same issues with M2 2.3.0, there are only approx 6K products and clean instance, no 3rd party modules. Luma Theme. Working in developer mode. 1 instance, three stores. The categories have anchors for attributes (filters).

Env:
dedicated rootserver
8 vCores
32 GB RAM
PHP 7.2.13
apache + nginx
